src/Service/Resolvers/MsSqlQueryBuilder.cs

```diff
@@ -110,28 +113,69 @@ public string Build(SqlExecuteStructure structure)
+            string queryToGetCountOfRecordWithPK = $"SELECT COUNT(*) as {COUNT_ROWS_WITH_GIVEN_PK} FROM {tableName} WHERE {pkPredicates}";
+
+            // Query to initiate transaction and get number of records with given PK.
+            string prefixQuery = $"SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;" +
+                $"BEGIN TRANSACTION;" +
+                $"DECLARE @ROWS_TO_UPDATE int;" +
+                $"SET @ROWS_TO_UPDATE = ({queryToGetCountOfRecordWithPK}); " +
+                $"{queryToGetCountOfRecordWithPK};";
+
+            // Final query to be executed for the given PUT/PATCH operation.
+            StringBuilder upsertQuery = new(prefixQuery);
+
+            // Query to update record (if there exists one for given PK).
+            StringBuilder updateQuery = new(
+                $"IF @ROWS_TO_UPDATE = 1" +
+                $"UPDATE {tableName} WITH(UPDLOCK) " +
+                $"SET {updateOperations} " +
+                $"OUTPUT {outputColumns} " +
+                $"WHERE {updatePredicates};");
+
+            // Append the update query to upsert query.
+            upsertQuery.Append(updateQuery);
+
+            if (!structure.IsFallbackToUpdate)
             {
-                return $"SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;BEGIN TRANSACTION; UPDATE {QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)} " +
-                    $"WITH(UPDLOCK) SET {Build(structure.UpdateOperations, ", ")} " +
-                    $"OUTPUT {MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted)} " +
-                    $"WHERE {predicates} " +
-                    $"IF @@ROWCOUNT = 0 " +
-                    $"BEGIN; " +
-                    $"INSERT INTO {QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)} ({Build(structure.InsertColumns)}) " +
-                    $"OUTPUT {MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted)} " +
-                    $"VALUES ({string.Join(", ", structure.Values)}) " +
-                    $"END; COMMIT TRANSACTION";
+                // Append the conditional to check if the insert query is to be executed or not.
+                // Insert is only attempted when no record exists corresponding to given PK.
+                upsertQuery.Append("ELSE ");
+
+                // Columns which are assigned some value in the PUT/PATCH request.
+                string insertColumns = Build(structure.InsertColumns);
+
+                // Predicates added by virtue of database policy for create operation.
+                string createPredicates = JoinPredicateStrings(structure.DbPolicyPredicatesForOperations[Config.Operation.Create]);
+
+                // Query to insert record (if there exists none for given PK).
+                StringBuilder insertQuery = new($"INSERT INTO {tableName} ({insertColumns}) OUTPUT {outputColumns}");
```

Comment:
>StringBuilder [](http://example.com/codeflow?start=16&length=13)

What benefits does stringbuilder give us here since there is a constant number of modifications to the string here (looks like three until end of method).

Reply:
Even though the num of concats are limited for a single query, the time complexity can quickly ramp up as we handle multiple requests. For reference: https://learn.microsoft.com/en-us/troubleshoot/developer/visualstudio/csharp/language-compilers/string-concatenation.

Reply:
aside: this remarks section offers more context when considering string vs stringbuilder: [https://learn.microsoft.com/dotnet/api/system.text.stringbuilder?view=net-6.0#the-string-and-stringbuilder-types](https://learn.microsoft.com/dotnet/api/system.text.stringbuilder?view=net-6.0#the-string-and-stringbuilder-types)